ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatRecord.java

```diff
@@ -1634,6 +1637,44 @@ public static ByteBuffer deserializeUserMetadataRecord(CrcInputStream crcStream)
     }
   }
 
+  /**
+   * Creating a {@link ByteBufferInputStream} from the {@link CrcInputStream} by sharing the underlying memory if the
+   * crcStream is built upon a {@link com.github.ambry.commons.CommonUtils.ByteBufferDataInputStream}.
+   * @param crcStream The crcStream to read {@link ByteBuffer} out.
+   * @param dataSize The size of {@link ByteBuffer}.
+   * @return The {@link ByteBufferInputStream}
+   * @throws IOException Unexpected IO errors.
+   */
+  static ByteBufferInputStream getByteBufferInputStreamForBlobRecord(CrcInputStream crcStream, int dataSize)
+      throws IOException {
+    ByteBufferInputStream output = null;
+    InputStream inputStream = crcStream.getUnderlyingInputStream();
+    if (inputStream instanceof CommonUtils.ByteBufferDataInputStream) {
+      ByteBuffer byteBuffer = ((CommonUtils.ByteBufferDataInputStream) inputStream).getBuffer();
+      int startPosition = byteBuffer.position();
+      // exhaust dataSize bytes and update crc for it.
+      byte[] dummy = new byte[4];
+      int read = 0;
+      while (read < (int) dataSize) {
+        int toRead = Math.min(4, (int) dataSize - read);
+        int n = crcStream.read(dummy, 0, toRead);
+        read += n;
+      }
+
+      int endPosition = byteBuffer.position();
+      int endLimit = byteBuffer.limit();
+      byteBuffer.position(startPosition);
+      byteBuffer.limit(endPosition);
+      output = new ByteBufferInputStream(byteBuffer.slice());
+      // resume the byteBuffer for future use.
+      byteBuffer.position(endPosition);
+      byteBuffer.limit(endLimit);
+    } else {
+      output = new ByteBufferInputStream(crcStream, (int) dataSize);
```

Comment:
it also seems that the various casts to int are not needed since dataSize is already an int.